jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseServiceImpl.java

```diff
@@ -0,0 +1,477 @@
+                caseInstance.setCaseMilestones(milestones);
+            }
+            if (withRoles) {
+                Collection<CaseRoleInstance> roles = getCaseRoleAssignments(caseId);
+                caseInstance.setCaseRoles(roles);
+            }
+            if (withStages) {
+                Collection<CaseStageInstance> stages = caseRuntimeDataService.getCaseInstanceStages(caseId, true, new org.kie.internal.query.QueryContext(0, 100));
+                caseInstance.setCaseStages(stages);
+            }
+            
+            return caseInstance;
+        } else {
+            return null;
+        }
+    }    
+
+    @Override
+    public void cancelCase(String caseId) throws CaseNotFoundException {
+        logger.debug("About to abort case {}", caseId);        
+        ProcessInstanceDesc pi = verifyCaseIdExists(caseId);        
+        processService.execute(pi.getDeploymentId(), CaseContext.get(caseId), new CancelCaseCommand(caseId, processService, runtimeDataService, false));
+    }
+
+    @Override
+    public void destroyCase(String caseId) throws CaseNotFoundException {
+        logger.debug("About to destroy permanently case {}", caseId);  
+        ProcessInstanceDesc pi = verifyCaseIdExists(caseId);        
+        processService.execute(pi.getDeploymentId(), CaseContext.get(caseId), new CancelCaseCommand(caseId, processService, runtimeDataService, true));
+    }
+    
+    /*
+     * Dynamic operations on a case
+     */
+    
+
+    @Override
+    public void addDynamicTask(String caseId, TaskSpecification taskSpecification) throws CaseNotFoundException {
+        ProcessInstanceDesc pi = verifyCaseIdExists(caseId);
+        
+        processService.execute(pi.getDeploymentId(), CaseContext.get(caseId), new AddDynamicTaskCommand(caseId, taskSpecification.getNodeType(), pi.getId(), taskSpecification.getParameters()));
+    }
+    
+    @Override
+    public void addDynamicTask(Long processInstanceId, TaskSpecification taskSpecification) throws ProcessInstanceNotFoundException {
+        ProcessInstanceDesc pi = runtimeDataService.getProcessInstanceById(processInstanceId);
+        if (pi == null || !pi.getState().equals(ProcessInstance.STATE_ACTIVE)) {
+            throw new ProcessInstanceNotFoundException("No process instance found with id " + processInstanceId + " or it's not active anymore");
+        }
+        String caseId = pi.getCorrelationKey();
+        processService.execute(pi.getDeploymentId(), ProcessInstanceIdContext.get(processInstanceId), new AddDynamicTaskCommand(caseId, taskSpecification.getNodeType(), pi.getId(), taskSpecification.getParameters()));
+    }
+
+    @Override
+    public void addDynamicTaskToStage(String caseId, String stageId, TaskSpecification taskSpecification) throws CaseNotFoundException, StageNotFoundException {
+        ProcessInstanceDesc pi = verifyCaseIdExists(caseId);
+
```

Comment:
I guess this would also need to check if the process instance exists and is active as other methods.